File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Common/StackAxis.kt
package com.abracode.actionui.Common

import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.compositionLocalOf

/**
 * The layout axis of the nearest enclosing stack, exposed to descendant
 * elements that must adapt to their container's orientation.
 *
 * SwiftUI infers some elements' orientation from context — most notably
 * `Divider`, which renders as a horizontal rule inside a `VStack` and a
 * vertical rule inside an `HStack`. Compose has no such inference: it ships
 * distinct `HorizontalDivider` / `VerticalDivider` composables, and a child has
 * no built-in way to learn whether it sits in a `Row` or a `Column`. This
 * CompositionLocal carries that fact down the tree so the orientation-sensitive
 * builders can mirror SwiftUI's automatic behavior.
 *
 * [VStack] provides [Vertical]; [HStack] provides [Horizontal]. The root
 * default is [Vertical] so a stack-less top-level `Divider` renders as a
 * horizontal rule, matching SwiftUI's default.
 */
enum class StackAxis {
    /** A vertical stack (Column / VStack): children are laid out top-to-bottom. */
    Vertical,
    /** A horizontal stack (Row / HStack): children are laid out start-to-end. */
    Horizontal
}

/**
 * CompositionLocal carrying the active [StackAxis]. Defaults to
 * [StackAxis.Vertical] (the SwiftUI default for a divider with no enclosing
 * stack). Uses [compositionLocalOf] — not the `static` variant — because the
 * value legitimately changes at every nested stack boundary, so subtrees that
 * read it must recompose when it does.
 */
val LocalStackAxis: ProvidableCompositionLocal<StackAxis> =
    compositionLocalOf { StackAxis.Vertical }

File: ActionUIWeb/src/Common/StackAxis.js
// StackAxis.js — shared stack layout vocabulary.
// Web analog of ActionUI/Common/StackAxis.kt (Android).
//
// SwiftUI infers some elements' orientation from their enclosing stack — most
// notably Divider, which is a horizontal rule inside a VStack and a vertical
// rule inside an HStack. The DOM has no such inference, so a stack tags its
// root with `data-aui-axis` and orientation-sensitive children (Divider) read
// the nearest ancestor's axis. The root default is "vertical", so a stack-less
// top-level Divider renders as a horizontal rule, matching SwiftUI.

export const StackAxis = Object.freeze({
    Vertical: "vertical",
    Horizontal: "horizontal",
});

// Cross-axis alignment maps. SwiftUI's VStack aligns children horizontally;
// its HStack aligns them vertically. Both default to center (matching SwiftUI,
// whose stack default is .center even though flexbox/Compose default elsewhere).
export const COLUMN_ALIGN = { leading: "flex-start", center: "center", trailing: "flex-end" };
export const ROW_ALIGN = {
    top: "flex-start", center: "center", bottom: "flex-end",
    firstTextBaseline: "baseline", lastTextBaseline: "last baseline",
};

export const DEFAULT_SPACING = 8; // approximates SwiftUI's adaptive default spacing

// Resolves the active axis for an orientation-sensitive child by walking up to
// the nearest stack root. Defaults to Vertical when there is no enclosing stack.
export function nearestStackAxis(node) {
    const stack = node.closest?.("[data-aui-axis]");
    return stack?.dataset.auiAxis ?? StackAxis.Vertical;
}
